typing/printtyp.ml

```diff
@@ -1638,19 +1638,37 @@ let rec tree_of_modtype ?(ellipsis=false) = function
       Omty_alias (tree_of_path Module p)
 
 and tree_of_signature sg =
-  wrap_env (fun env -> env) (tree_of_signature_rec !printing_env false) sg
-
-and tree_of_signature_rec env' in_type_group = function
-    [] -> stop_type_group env'; []
-  | item :: rem as items ->
-      let in_type_group = still_in_type_group env' in_type_group item in
-      let (sg, rem) = filter_rem_sig item rem in
-      hide_rec_items items;
-      protect_rec_items items;
-      reset_naming_context ();
-      let trees = trees_of_sigitem item in
-      let env' = Env.add_signature (item :: sg) env' in
-      trees @ tree_of_signature_rec env' in_type_group rem
+  wrap_env (fun env -> env)
+    (tree_of_signature_rec (fun _ (_,x) -> x) !printing_env) sg
+
+and tree_of_signature_rec : 'r. (_->_->'r) -> _ -> _ -> 'r list
+  = fun decorate env' sg ->
+  let structured = group_recursive_items (group_syntactic_items sg) in
+  let collect_trees_of_rec_group (env,trees) group =
+    let env', more_trees = trees_of_recursive_sigitem_group env group in
```

Comment:
I find the name `more_trees` confusing. Initially I thought that this was a recursive function that would return an output for some part of the input it consumed, and then rest of the input. I think that `group_trees` would be a better name. 

Note:  with the no-env-passing suggestion you could also get rid of the fold style, use `List.concat_map`, and thus just call it `trees` if you wished.

Reply:
Renamed to group_trees.